docs(core): add documentation about custom steps for nx cloud workflows (#23385)
nrwl/nx: docs/nx-cloud/reference/custom-steps.md

# Creating Custom Steps

While Nx Cloud provides several pre-built steps, you can create custom steps to suit your specific needs. Custom steps are defined by creating a new YAML file.

## Defining a Custom Step

To create a custom step, follow these steps:

1. **Create a New YAML File**: Create a new YAML file in your `.nx/workflows` directory (e.g., `custom-step.yaml`)

{% callout type="note" title="Custom step file location" %}
Custom steps do not need to be in the `.nx/workflows` directory. However, they must be available on a public GitHub repository. You have a few options for organizing your custom steps:

- **Create a Dedicated Repository**: You can create a separate repository that contains a collection of custom steps. This can be useful for sharing steps across multiple projects or teams.

> - **Same Repository as your Nx Workspace**: You can place the custom step file in the same repository as your Nx workspace. Ensure that the custom step file is publicly accessible on GitHub.

Reviewer: ```suggestion
- **Same Repository as your Nx Workspace**: You can place the custom step file in the same repository as your Nx workspace. Ensure that the custom step file is publicly accessible on GitHub. If your project is not on GitHub you will need to create a dedicate repo for your steps on GitHub.
```

  - **steps**: A list of sub-steps to be executed (only if `using` is `aggregate`). Steps follow the same definition as [launch templates](/ci/reference/launch-templates#launch-template-structure).

- **inputs**:
  - **name**: The name of the input.
  - **description**: A description of the input.
  - **default**: Default value for the input.
  - **required**: Whether the input is required.

### Using Inputs in Scripts

If your custom step has inputs, they can be accessed in scripts or JavaScript files using environment variables. Each input is prefixed with `NX_CLOUD_INPUT_`. For example, if you have an input called `input1`, you can access it in a JavaScript file like this:

```javascript
const input1 = process.env.NX_CLOUD_INPUT_input1;
console.log(`The value of input1 is: ${input1}`);
```

### Using Custom Steps in Launch Templates

Once you've defined your custom step, you can use it in your launch templates by referencing the custom step file. Here's an example:

```yaml {% fileName=".nx/workflows/agents.yaml" %}
launch-templates:
  custom-template:
    resource-class: 'docker_linux_amd64/medium'
    image: 'ubuntu22.04-node20.11-v7'
    init-steps:
      - name: Custom Step

>         uses: 'your-org/custom-steps-repo/.nx/workflows/custom-steps.yaml'

Reviewer: ```suggestion
        uses: 'your-org/your-repo/main/.nx/workflows/custom-steps.yaml'
```

Reviewer: they would need a branch I guess?

Reviewer: I also changed to use `your-repo` since we're recommending people add this to their existing repo.

>         uses: 'your-org/custom-steps-repo/.nx/workflows/custom-steps.yaml'

Reviewer: The above won't work for private Github repos. Is it worth do you think adding an example of:

```yaml
https://x-access-token:{{secrets.PERSONAL_ACCESS_TOKEN}}@raw.githubusercontent.com/your-org/your-repo/main/.nx/workflows/custom-steps.yaml
```

Or do you think it'll confuse everyone more?

Author: We'll add more details for this at a later time until other things are resolved

        env:

>           CUSTOM_VAR: 'custom_value'

Reviewer: it might be better to not give an example of `env:` as it might confuse people on how it's different from inputs

`env` is not that useful anymore now that we have inputs?

Author: I think having env shown is good, because there could be cases where someone just wants a regular env without having to deal with `NX_CLOUD_INPUT_`. I can add more details that using inputs would enable better validation though.

        inputs:
          input1: 'value1'
          input2: 'value2'
```

### Validating Custom Steps

> Just like launch templates, you should validate your custom steps to ensure they are defined correctly. Use the `nx-cloud validate` command with the `--stepFile` flag to do this:

Reviewer: ```suggestion
Just like launch templates, you should validate your custom steps to ensure they are defined correctly. In your CI pipeline, use the `nx-cloud validate` command with the `--stepFile` flag to ensure nobody merges in invalid changes to your step:
```
